### PR TITLE
Change ORT/atstccfg to generate all files at once, massive performance improvement

### LINKED ISSUE #4436: ORT uses non-standard API route /ort/{{host name}}/packages
## I'm submitting a ...
- improvement request (usability, performance, tech debt, etc.)

## Traffic Control components affected ...
- CDN-in-a-Box (because of ORT.py)
- Traffic Ops ORT (and also the Python version used by CDN-in-a-Box)

## Current behavior:
To fetch the packages necessary for install on cache servers, ORT calls the `/ort/{{host name}}/packages` endpoint, which is not grounded in an API version nor documented anywhere, nor handled in Go.

## Expected / new behavior:
ORT should use only real, documented, supported, versioned, API endpoints. I'm not personally very particular how that's done, though I prefer the bottom-most option (unrealistic though that may be).

• The call could be forwarded to ATSTCCFG which could be updated to support generating the output from standard API calls
• ORT itself could be changed to request the necessary information from the standard API
• The endpoint could be rewritten in Go and added to the 2.0 API.
• The ability to install packages could be removed from ORT entirely, and be left to the orchestration or administration system of choice for the organization managing the server - which probably does it better anyway.

## Minimal reproduction of the problem with instructions:
Run ORT, observe the call to an unsupported, non-standard API

### LINKED ISSUE #4437: ORT uses non-standard API endpoint /ort/{{host name}}/chkconfig
## I'm submitting a ...
- improvement request (usability, performance, tech debt, etc.)

## Traffic Control components affected ...
- CDN-in-a-Box (because of ORT.py)
- Traffic Ops ORT (and also the Python version used by CDN-in-a-Box)

## Current behavior:
To fetch the services and their necessary states in different run-levels on cache servers, ORT calls the `/ort/{{host name}}/chkconfig` endpoint, which is not grounded in an API version nor documented anywhere, nor handled in Go.

## Expected / new behavior:
ORT should use only real, documented, supported, versioned, API endpoints. I'm not personally very particular how that's done, though I prefer the bottom-most option (unrealistic though that may be).

• The call could be forwarded to ATSTCCFG which could be updated to support generating the output from standard API calls
• ORT itself could be changed to request the necessary information from the standard API
• The endpoint could be rewritten in Go and added to the 2.0 API.
• The ability to manage services that are not ATS could be removed from ORT entirely and left up to the orchestration or administration system in use by the organization managing the cache server, as they probably do it much better.

## Minimal reproduction of the problem with instructions:
Run ORT, observe the call to an unsupported, non-standard API

## Anything Else
However this is done - provided the functionality is retained - it really ought to not use the 'chkconfig' format, as all CentOS versions currently supported use SystemD instead.

### LINKED ISSUE #4488: atstccfg logs incorrect message with --no-cache
## I'm submitting a ...
- bug report

## Traffic Control components affected ...
- Traffic Ops ORT (`atstccfg`)

## Current behavior:
If you try to request a file through ATSTCCFG with caching disabled, it will tell you the cached files don't exist instead of that it skipped checking them.

## Expected / new behavior:
ATSTCCFG should log accurate messages, so that someone reading them later won't spend any time wondering about permissions or scouring code for typos. Not that I would know anything about that.

## Minimal reproduction of the problem with instructions:
Do this twice and note the output (once isn't sufficient to display the problem since you may actually just not have a cache if it's your first time running it).
```bash
/path/to/atstccfg -s -P "$TO_PASSWORD" -U "$TO_USER" -u "$TO_URL/api/2.0/servers/mid/configfiles/ats/remap.config"
```

## Anything else:
<details>
<summary>Sample run with CDN-in-a-Box</summary>
<pre>
$ ./atstccfg --no-cache  -s -P twelve -U admin -u "https://127.0.0.1:6443/api/2.0/servers/mid/configfiles/ats/remap.config"
INFO: atstccfg.go:83: 2020-03-12T18:54:26.403013307-06:00: URL: 'https://127.0.0.1:6443/api/2.0/servers/mid/configfiles/ats/remap.config' User: 'admin' Pass len: '6'
INFO: atstccfg.go:84: 2020-03-12T18:54:26.403057237-06:00: TempDir: '/tmp/atstccfg_cache'
INFO: atstccfg.go:87: 2020-03-12T18:54:26.40306432-06:00: TO FQDN: 'https://127.0.0.1:6443'
INFO: atstccfg.go:88: 2020-03-12T18:54:26.403074515-06:00: TO URL: 'https://127.0.0.1:6443/api/2.0/servers/mid/configfiles/ats/remap.config'
INFO: trafficops.go:50: 2020-03-12T18:54:26.40309095-06:00: failed to get cookies from cache file (trying real TO): opening cookie file '/tmp/atstccfg_cache/cookies':open /tmp/atstccfg_cache/cookies: no such file or directory
INFO: trafficops.go:60: 2020-03-12T18:54:26.451591885-06:00: using cookies from TO
INFO: routing.go:73: 2020-03-12T18:54:26.451638922-06:00: GetConfigFile scope 'servers' resource 'mid' fileName 'remap.config'
INFO: routing.go:205: 2020-03-12T18:54:26.451645392-06:00: GetConfigFileServer server 'mid' fileName 'remap.config'
INFO: caching.go:68: 2020-03-12T18:54:26.45166684-06:00: ReadCache failed to get object from '/tmp/atstccfg_cache/servers.json', calling getter: opening object file '/tmp/atstccfg_cache/servers.json':open /tmp/atstccfg_cache/servers.json: no such file or directory
INFO: caching.go:91: 2020-03-12T18:54:26.468867806-06:00: GetCachedJSON servers.json (miss) took 17ms
INFO: caching.go:68: 2020-03-12T18:54:26.468899554-06:00: ReadCache failed to get object from '/tmp/atstccfg_cache/cdn_CDN-in-a-Box.json', calling getter: opening object file '/tmp/atstccfg_cache/cdn_CDN-in-a-Box.json':open /tmp/atstccfg_cache/cdn_CDN-in-a-Box.json: no such file or directory
INFO: caching.go:91: 2020-03-12T18:54:26.470840223-06:00: GetCachedJSON cdn_CDN-in-a-Box.json (miss) took 2ms
INFO: caching.go:68: 2020-03-12T18:54:26.470861255-06:00: ReadCache failed to get object from '/tmp/atstccfg_cache/profile_global_parameters.json', calling getter: opening object file '/tmp/atstccfg_cache/profile_global_parameters.json':open /tmp/atstccfg_cache/profile_global_parameters.json: no such file or directory
INFO: caching.go:91: 2020-03-12T18:54:26.473073498-06:00: GetCachedJSON profile_global_parameters.json (miss) took 2ms
INFO: caching.go:68: 2020-03-12T18:54:26.473098588-06:00: ReadCache failed to get object from '/tmp/atstccfg_cache/profile_ATS_MID_TIER_CACHE_parameters.json', calling getter: opening object file '/tmp/atstccfg_cache/profile_ATS_MID_TIER_CACHE_parameters.json':open /tmp/atstccfg_cache/profile_ATS_MID_TIER_CACHE_parameters.json: no such file or directory
INFO: caching.go:91: 2020-03-12T18:54:26.476085698-06:00: GetCachedJSON profile_ATS_MID_TIER_CACHE_parameters.json (miss) took 3ms
INFO: caching.go:68: 2020-03-12T18:54:26.476113523-06:00: ReadCache failed to get object from '/tmp/atstccfg_cache/cdn_2_deliveryservices.json', calling getter: opening object file '/tmp/atstccfg_cache/cdn_2_deliveryservices.json':open /tmp/atstccfg_cache/cdn_2_deliveryservices.json: no such file or directory
INFO: caching.go:91: 2020-03-12T18:54:26.482338274-06:00: GetCachedJSON cdn_2_deliveryservices.json (miss) took 6ms
INFO: caching.go:68: 2020-03-12T18:54:26.482364366-06:00: ReadCache failed to get object from '/tmp/atstccfg_cache/deliveryservice_servers_s_d_.csv', calling getter: opening object file '/tmp/atstccfg_cache/deliveryservice_servers_s_d_.csv':open /tmp/atstccfg_cache/deliveryservice_servers_s_d_.csv: no such file or directory
INFO: caching.go:91: 2020-03-12T18:54:26.48529477-06:00: GetCachedJSON deliveryservice_servers_s_d_.csv (miss) took 3ms
INFO: caching.go:68: 2020-03-12T18:54:26.485336655-06:00: ReadCache failed to get object from '/tmp/atstccfg_cache/ds_regexes.json', calling getter: opening object file '/tmp/atstccfg_cache/ds_regexes.json':open /tmp/atstccfg_cache/ds_regexes.json: no such file or directory
INFO: caching.go:91: 2020-03-12T18:54:26.490864101-06:00: GetCachedJSON ds_regexes.json (miss) took 6ms
INFO: caching.go:64: 2020-03-12T18:54:26.491237212-06:00: ReadCache profile_ATS_MID_TIER_CACHE_parameters.json (hit) took 0s
INFO: caching.go:68: 2020-03-12T18:54:26.491262735-06:00: ReadCache failed to get object from '/tmp/atstccfg_cache/config_file_cachekey.config_parameters.json', calling getter: opening object file '/tmp/atstccfg_cache/config_file_cachekey.config_parameters.json':open /tmp/atstccfg_cache/config_file_cachekey.config_parameters.json: no such file or directory
INFO: caching.go:91: 2020-03-12T18:54:26.493827369-06:00: GetCachedJSON config_file_cachekey.config_parameters.json (miss) took 3ms
INFO: caching.go:68: 2020-03-12T18:54:26.493867734-06:00: ReadCache failed to get object from '/tmp/atstccfg_cache/cachegroups.json', calling getter: opening object file '/tmp/atstccfg_cache/cachegroups.json':open /tmp/atstccfg_cache/cachegroups.json: no such file or directory
INFO: caching.go:91: 2020-03-12T18:54:26.496658951-06:00: GetCachedJSON cachegroups.json (miss) took 3ms
INFO: routing.go:43: 2020-03-12T18:54:26.49667983-06:00: GetConfigFile /api/2.0/servers/mid/configfiles/ats/remap.config took 45ms
INFO: atstccfg.go:104: 2020-03-12T18:54:26.496685342-06:00: GetConfigFile returned 0 <nil>
# DO NOT EDIT - Generated for mid by Traffic Ops (https://trafficops.infra.ciab.test:443/) on Fri Mar 13 00:54:26 UTC 2020
map http://origin.infra.ciab.test http://origin.infra.ciab.test @plugin=cache_range_requests.so
$ ./atstccfg --no-cache  -s -P twelve -U admin -u "https://127.0.0.1:6443/api/2.0/servers/mid/configfiles/ats/remap.config"
INFO: atstccfg.go:83: 2020-03-12T18:54:26.403013307-06:00: URL: 'https://127.0.0.1:6443/api/2.0/servers/mid/configfiles/ats/remap.config' User: 'admin' Pass len: '6'
INFO: atstccfg.go:84: 2020-03-12T18:54:26.403057237-06:00: TempDir: '/tmp/atstccfg_cache'
INFO: atstccfg.go:87: 2020-03-12T18:54:26.40306432-06:00: TO FQDN: 'https://127.0.0.1:6443'
INFO: atstccfg.go:88: 2020-03-12T18:54:26.403074515-06:00: TO URL: 'https://127.0.0.1:6443/api/2.0/servers/mid/configfiles/ats/remap.config'
INFO: trafficops.go:50: 2020-03-12T18:54:26.40309095-06:00: failed to get cookies from cache file (trying real TO): opening cookie file '/tmp/atstccfg_cache/cookies':open /tmp/atstccfg_cache/cookies: no such file or directory
INFO: trafficops.go:60: 2020-03-12T18:54:26.451591885-06:00: using cookies from TO
INFO: routing.go:73: 2020-03-12T18:54:26.451638922-06:00: GetConfigFile scope 'servers' resource 'mid' fileName 'remap.config'
INFO: routing.go:205: 2020-03-12T18:54:26.451645392-06:00: GetConfigFileServer server 'mid' fileName 'remap.config'
INFO: caching.go:68: 2020-03-12T18:54:26.45166684-06:00: ReadCache failed to get object from '/tmp/atstccfg_cache/servers.json', calling getter: opening object file '/tmp/atstccfg_cache/servers.json':open /tmp/atstccfg_cache/servers.json: no such file or directory
INFO: caching.go:91: 2020-03-12T18:54:26.468867806-06:00: GetCachedJSON servers.json (miss) took 17ms
INFO: caching.go:68: 2020-03-12T18:54:26.468899554-06:00: ReadCache failed to get object from '/tmp/atstccfg_cache/cdn_CDN-in-a-Box.json', calling getter: opening object file '/tmp/atstccfg_cache/cdn_CDN-in-a-Box.json':open /tmp/atstccfg_cache/cdn_CDN-in-a-Box.json: no such file or directory
INFO: caching.go:91: 2020-03-12T18:54:26.470840223-06:00: GetCachedJSON cdn_CDN-in-a-Box.json (miss) took 2ms
INFO: caching.go:68: 2020-03-12T18:54:26.470861255-06:00: ReadCache failed to get object from '/tmp/atstccfg_cache/profile_global_parameters.json', calling getter: opening object file '/tmp/atstccfg_cache/profile_global_parameters.json':open /tmp/atstccfg_cache/profile_global_parameters.json: no such file or directory
INFO: caching.go:91: 2020-03-12T18:54:26.473073498-06:00: GetCachedJSON profile_global_parameters.json (miss) took 2ms
INFO: caching.go:68: 2020-03-12T18:54:26.473098588-06:00: ReadCache failed to get object from '/tmp/atstccfg_cache/profile_ATS_MID_TIER_CACHE_parameters.json', calling getter: opening object file '/tmp/atstccfg_cache/profile_ATS_MID_TIER_CACHE_parameters.json':open /tmp/atstccfg_cache/profile_ATS_MID_TIER_CACHE_parameters.json: no such file or directory
INFO: caching.go:91: 2020-03-12T18:54:26.476085698-06:00: GetCachedJSON profile_ATS_MID_TIER_CACHE_parameters.json (miss) took 3ms
INFO: caching.go:68: 2020-03-12T18:54:26.476113523-06:00: ReadCache failed to get object from '/tmp/atstccfg_cache/cdn_2_deliveryservices.json', calling getter: opening object file '/tmp/atstccfg_cache/cdn_2_deliveryservices.json':open /tmp/atstccfg_cache/cdn_2_deliveryservices.json: no such file or directory
INFO: caching.go:91: 2020-03-12T18:54:26.482338274-06:00: GetCachedJSON cdn_2_deliveryservices.json (miss) took 6ms
INFO: caching.go:68: 2020-03-12T18:54:26.482364366-06:00: ReadCache failed to get object from '/tmp/atstccfg_cache/deliveryservice_servers_s_d_.csv', calling getter: opening object file '/tmp/atstccfg_cache/deliveryservice_servers_s_d_.csv':open /tmp/atstccfg_cache/deliveryservice_servers_s_d_.csv: no such file or directory
INFO: caching.go:91: 2020-03-12T18:54:26.48529477-06:00: GetCachedJSON deliveryservice_servers_s_d_.csv (miss) took 3ms
INFO: caching.go:68: 2020-03-12T18:54:26.485336655-06:00: ReadCache failed to get object from '/tmp/atstccfg_cache/ds_regexes.json', calling getter: opening object file '/tmp/atstccfg_cache/ds_regexes.json':open /tmp/atstccfg_cache/ds_regexes.json: no such file or directory
INFO: caching.go:91: 2020-03-12T18:54:26.490864101-06:00: GetCachedJSON ds_regexes.json (miss) took 6ms
INFO: caching.go:64: 2020-03-12T18:54:26.491237212-06:00: ReadCache profile_ATS_MID_TIER_CACHE_parameters.json (hit) took 0s
INFO: caching.go:68: 2020-03-12T18:54:26.491262735-06:00: ReadCache failed to get object from '/tmp/atstccfg_cache/config_file_cachekey.config_parameters.json', calling getter: opening object file '/tmp/atstccfg_cache/config_file_cachekey.config_parameters.json':open /tmp/atstccfg_cache/config_file_cachekey.config_parameters.json: no such file or directory
INFO: caching.go:91: 2020-03-12T18:54:26.493827369-06:00: GetCachedJSON config_file_cachekey.config_parameters.json (miss) took 3ms
INFO: caching.go:68: 2020-03-12T18:54:26.493867734-06:00: ReadCache failed to get object from '/tmp/atstccfg_cache/cachegroups.json', calling getter: opening object file '/tmp/atstccfg_cache/cachegroups.json':open /tmp/atstccfg_cache/cachegroups.json: no such file or directory
INFO: caching.go:91: 2020-03-12T18:54:26.496658951-06:00: GetCachedJSON cachegroups.json (miss) took 3ms
INFO: routing.go:43: 2020-03-12T18:54:26.49667983-06:00: GetConfigFile /api/2.0/servers/mid/configfiles/ats/remap.config took 45ms
INFO: atstccfg.go:104: 2020-03-12T18:54:26.496685342-06:00: GetConfigFile returned 0 <nil>
# DO NOT EDIT - Generated for mid by Traffic Ops (https://trafficops.infra.ciab.test:443/) on Fri Mar 13 00:54:26 UTC 2020
map http://origin.infra.ciab.test http://origin.infra.ciab.test @plugin=cache_range_requests.so
</pre>
</details>

### PR DESCRIPTION
**Pull request by rob05c**
## What does this Pull Request (PR) do?

Changes ORT/atstccfg to generate all files at once. ORT calls `atstccfg` once, and it returns `multipart/mixed` text of all files. ORT then parses the multipart into its variables, and processes all files as normal.

This reduces an ORT run for a large CDN from about 8 minutes to about 20 seconds. Fixing https://github.com/apache/trafficcontrol/issues/4473 brings it down to about 10 seconds.

This also:
- includes fixes to a number of ORT slugs I found. 
- eliminates atstccfg caching, because it's no longer necessary, as atstccfg is only called once for the config file data, and in turn only calls Traffic Ops once for those API endpoints.
- adds data arguments to atstccfg for non-config data got and put by ORT, and changes ORT to never directly call TO, but to go through atstccfg for all TO data.
- Makes atstccfg build the non-API packages and chkconfig endpoints from API data, so it's not necessary to create new API endpoints for those. Now, the only non-API endpoint used by ORT/atstccfg is `/update`, whose behavior cannot be replicated with existing API endpoints.
- adds a missing client function for an API endpoint needed, `GetServerUpdateStatus`.
- moves preprocessing, e.g. `__RETURN__` replacement, from ORT to atstccfg. This was necessary, because ORT no longer has the metadata necessary.

These changes should be transparent to ORT users. ORT itself doesn't change (except to remove the now-unused cache-related flags).

Includes tests for new atstccfg funcs and the client func it adds. Changes are also covered by exitsing config tests.
Includes documentation for preexisting ORT behavior. Though it doesn't actually change the interface.
No changelog, no interface change.

- [x] This PR fixes #4436 and fixes #4437 and fixes #4488

## Which Traffic Control components are affected by this PR?
- Traffic Ops ORT

## What is the best way to verify this PR?
Run tests.
Run ORT, verify it still generates configs properly as before.

## If this is a bug fix, what versions of Traffic Control are affected?
Not a bug fix.

## The following criteria are ALL met by this PR
- [x] This PR includes tests OR I have explained why tests are unnecessary
- [x] This PR includes documentation OR I have explained why documentation is unnecessary
- [x] This PR includes an update to CHANGELOG.md OR such an update is not necessary
- [x] This PR includes any and all required license headers
- [x] This PR ensures that database migration sequence is correct OR this PR does not include a database migration
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://www.apache.org/security/) for details)

## Additional Information

### CODE REVIEW

**Comment by rob05c** [comment]
retest this please

**Comment by ocket8888** [comment]
slugs = bugs?

**Comment by rob05c** [comment]
>includes fixes to a number of ORT slugs I found.

If anyone wants more detail: the big ones were `check_this_plugin` and `package_installed`. These funcs call OS yum data about installed packages, which is reasonably slow. They were calling out for every single plugin on every single remap line, ~2000 calls on a large CDN. But we only have ~10 unique plugins and packages.

I changed them to cache results, so they only ever call out to Yum/CentOS once per unique name.

That saved about 2 minutes on my large CDN run, reducing the full ORT run (after the single-file changes and other performance fixes) from about 140 seconds to about 20 seconds.

File: traffic_ops/ort/traffic_ops_ort.pl [hunk 1]
```diff
@@ -1833,159 +1558,103 @@ sub check_log_level {
...
+
+		my $file_name = basename($headers{'path'});
+		if ( length($file_name) == 0 ) {
+			print "FATAL Error getting package list from Traffic Ops! Headers on file $i missing Path!\n";
+			exit 1;
 		}
+
+		my %file_obj;
+		$file_obj{'headers'} = \%headers;
+		$file_obj{'body'} = $headers_body[1];
```
[the remark this answers is not in the document]
Comment (traeak) [reply traffic_ops/ort/traffic_ops_ort.pl]: oops invalid perl comment.

[the remark this answers is not in the document]
Comment (rob05c) [reply traffic_ops/ort/traffic_ops_ort.pl]: Fixed.
I went ahead and made it trim all whitespace and then add the requisite \n. It doesn't seem noticeably slower, and it's just a little safer.

[the remark this answers is not in the document]
Comment (traeak) [reply traffic_ops/ort/traffic_ops_ort.pl]: Not totally happy with this fix but it's practical enough for now until ort is rewritten in go which has library support for multi part http bodies.

**Pull request merged**